Source/opal/OpalBridge.m

```diff
@@ -0,0 +1,126 @@
+/*
+   OpalBridge.m
+
+   Copyright (C) 2017 Free Software Foundation, Inc.
+
+   Author: Daniel Ferreira <dtf@stanford.edu>
+   Date: July 2017
+
+   This file is part of GNUstep.
+
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2 of the License, or (at your option) any later version.
+
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; see the file COPYING.LIB.
+   If not, see <http://www.gnu.org/licenses/> or write to the
+   Free Software Foundation, 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110-1301, USA.
+*/
+
+#import <AppKit/NSColor.h>
+#import <AppKit/NSImage.h>
+#import <AppKit/NSColorSpace.h>
+#import <AppKit/NSGraphics.h>
+#import <CoreGraphics/CGColor.h>
+#import <CoreGraphics/CGImage.h>
+#import <objc/runtime.h>
+
+@interface GSNamedColor : NSColor
+- (NSColor *)cachedColor;
+@end
+@interface GSWhiteColor : NSColor
+@end
+@interface GSDeviceCMYKColor : NSColor
+@end
+@interface GSRGBColor : NSColor
+@end
+@interface GSPatternColor : NSColor
+@end
+
+@implementation NSColor (GSQuartz)
+/*
+ * FIXME:
+ * NOTE 1: GNUstep-GUI does not allow an NSColor to be created with a custom
+ * NSColorSpace. If this were allowed, we'd have to:
+ *   1) implement a bridge for -[NSColorSpace CGColorSpace]
+ *   2) for each color, extract that color space and generate a CGColorRef.
+ * NOTE 2: GNUstep-GUI makes no distinction of device and generic color spaces.
+ * If this ever ceases to be the case, some adjustment might be necessary here.
+ */
+- (CGColorRef)CGColor
+{
+  NSString *name = [self colorSpaceName];
+  CFStringRef cgColorSpaceName = NULL;
+  if ([name isEqualToString: NSCalibratedRGBColorSpace]
+    || [name isEqualToString: NSDeviceRGBColorSpace])
+    cgColorSpaceName = kCGColorSpaceSRGB;
+  else if ([name isEqualToString: NSCalibratedBlackColorSpace]
+    || [name isEqualToString: NSCalibratedWhiteColorSpace]
+    || [name isEqualToString: NSDeviceBlackColorSpace]
+    || [name isEqualToString: NSDeviceWhiteColorSpace])
+    cgColorSpaceName = kCGColorSpaceGenericGray;
+  else if ([name isEqualToString: NSDeviceCMYKColorSpace])
+    cgColorSpaceName = kCGColorSpaceGenericCMYK;
+  else if ([name isEqualToString: NSNamedColorSpace])
+    {
+      // FIXME
+    }
+
+  if (cgColorSpaceName == NULL)
+    return NULL;
+
+  CGColorSpaceRef colorSpace = CGColorSpaceCreateWithName(cgColorSpaceName);
+  CGFloat values[10];
+  [self getComponents: values];
+
+  CGColorRef color = CGColorCreate(colorSpace, values);
```

Comment:
As I wrote in a mail, this is wrong for the NSCalibratedBlackColorSpace] and NSDeviceBlackColorSpace] colour spaces. I think it would be better to ignore them in the code above.